Add metadata API (#882)
measuredco/puck: apps/docs/pages/docs/api-reference/configuration/component-config.mdx

#### Render props

| Arg                                             | Example       | Type     |
| ----------------------------------------------- | ------------- | -------- |
| [`id`](#id)                                     | `button-1234` | String   |

> | [`puck.metadata`](/docs/api-reference/metadata) | `{}`          | Object   |

Reviewer: Can you make alphabetical please?

| [`puck.dragRef`](#puckdragref)                  | `null`        | Function |
| [`puck.isEditing`](#puckisediting)              | `false`       | Boolean  |
| [`puck.renderDropZone`](#puckrenderdropzone)    | `() => {}`    | Function |
| [`...props`](#props)                            | `{}`          | Object   |

##### `id`

A unique ID generated by Puck for this component. You can optionally apply this, or use your own ID.

> ##### `puck.metadata`

Reviewer: Can you make alphabetical please?

> Global metadata passed in to the `<Puck>` component. Can be used along with component props for rendering.
> Using metadata within the `render` API is non-destructive and will conserve the values saved in the component props.

Reviewer: Think this can be simplified

```suggestion
Global metadata that is be passed in via the Puck [`metadata` prop](/docs/api-reference/components/puck#metadata).
```

The data object from the previous run of this function.

##### `params.metadata`

Global metadata provided to the `<Puck>` component. See the [`Metadata` docs](/docs/api-reference/metadata) for a full reference.

```tsx copy {2-4} /changed/1 filename="Example only updating 'resolvedTitle' when 'title' changes"
const resolveData = async ({ props }, { metadata }) => {
  return {
    props: { title: metadata.title || props.title },
  };
};
```

Note that returning metadata in the resolveData API will overwrite the prop values. For non-destructive use of metadata, refer to the [`render(puck.metadata)` docs](#puckmetadata).

> 

Reviewer: Hm I hadn't thought that we might need to expose this in `resolveData`, but I think it makes sense 🤔

Author: Since metadata can be anything, a value contained within could be used in resolveData as a conditional operator. For example, `if (metadata.product.type === "water") return { props: { description: "It's very wet" } };`
